brokers/artifacts/extensions.go

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright (c) 2018-2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package artifacts
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"path/filepath"
+
+	"github.com/eclipse/che-plugin-broker/model"
+)
+
+const installedPluginsJSONFile = "/plugins/installed.json"
+const installedPluginsJSONVersion = "1.0"
+
+// SyncPlugins takes a list of requested plugins and resolves it against what is currently
+// installed, returning a list of InstalledPlugins that need to be downloaded. Note that
+// returned list may be partially filled, if e.g. only one of two extensions in a plugin
+// need to be updated.
+// Plugins that are out of date will be deleted from the local filesystem, and any extensions
+// that are not included in the current requested plugins will be removed.
+func (b *Broker) SyncPlugins(requested []model.CachedPlugin) []model.CachedPlugin {
+	installed, err := b.readInstalledPlugins()
+	if err != nil {
+		// Unable to read -- default to wiping everything and installing all plugins
+		b.PrintDebug("Error encountered during reading installed plugins: %s", err)
+		b.PrintInfo("Failed to get installed plugins")
+		b.resetPluginsDirectory()
+		return requested
+	}
+	toInstall := b.getPlan(requested, installed)
+	return toInstall
+}
+
+// getPlan will prepare the /plugins directory for installation and return a list of (partially-filled)
+// plugins to be installed. Any plugins that are currently installed but not requested in the workspace will
+// be removed from the filesystem.
+func (b *Broker) getPlan(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin) {
+	toInstall = requested
+	for _, plugin := range installed {
+		match := findPlugin(plugin, requested)
+		if match == nil {
+			// Plugin has been uninstalled since last start
+			b.PrintInfo("Uninstalling plugin: %s", plugin.ID)
+			if err := b.removePlugin(plugin); err != nil {
```

Comment:
It was quite unexpected for me that `getPlan` method actually removes plugins, maybe instead of performing any operation here it makes sense to keep it only analyzing extenstions and return plan:
```
func (b *Broker) getPlan(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin, toRemove []model.CachedPlugin, filesToRemove []File) {
```
and then remove them on another level? Or maybe there can be found a more clear method name.

P.S. I have not had reviewed your changes deeply enough if my comment does not make any sense - please just mark it resolved =)

Reply:
You're right that the name is not very good, I will improve that. The issue I ran into with trying to get a plan and then execute it is that there are three cases:
1. A plugin is added to the workspace -> download its extensions
2. A plugin is removed from the workspace -> remove its extensions
3. A plugin is updated in the workspace -> remove outdated extensions, download new ones

The third case is hard to support and also potentially the most common (`latest` represents an "update", sort of). What I opted for, as a result, is to partially fill plugins in the plan to support case three (if an extension has already been downloaded, it has the extension's path added to the struct).

Looking back this is probably over-engineered, since it only occurs when e.g. jdt is updated but the debug extension is not.

Reply:
What about scenarios where the same extension (but different versions) are installed because of plugins which depend on different versions? eg., 
```
redhat/java8/latest 
- https://download.jboss.org/jbosstools/vscode/3rdparty/vscode-java-debug/vscode-java-debug-0.20.0.vsix

redhat/quarkus-java8
- https://download.jboss.org/jbosstools/vscode/3rdparty/vscode-java-debug/vscode-java-debug-0.23.0.vsix
```

Do you make sure that an update for one plugin (eg., newer quarkus results in 0.23.0 -> 0.24.0) doesn't also garbage collect the 0.2.0 used by the other plugin?

Reply:
Currently, each plugin gets its own .vsix on its own path, so there's no overlap for that case. E.g. redhat/java plugins end up in
```
/plugins
└── sidecars
    └── redhat_java_0_43_0
        ├── redhat.java.0.43.0.kmwjfvvtpc.java-0.43.0-1473.vsix
        └── redhat.java.0.43.0.wgfmdnrsmo.vscode-java-debug-0.16.0.vsix
```